template/upstream.h.erb

```diff
@@ -16,10 +16,22 @@ typedef void <%= e.name_sanitized %>_callback_f(
 <%# The Function who register the callbacks to the dispatcher %>
 void btx_register_callbacks_<%= e.name_sanitized %>(
     void *btx_handle, <%= e.name_sanitized %>_callback_f *callback);
-<% end %>
+void _btx_register_callbacks_<%= e.name_sanitized %>(
+    void *btx_handle, const char *callback_type, void *callback);
 
+<% end %>
 void btx_delete_dispatchers(common_data_t *common_data);
 
+// Matching callbacks
+
+<% matching_dispatchers.each do |m| %>
+typedef void (<%= m.name_sanitized %>_callback_f)(
+    void *btx_handle,
+    void *usr_data<%= m.args.map{ |s| "#{s.type} #{s.name}" }.join_with_prefix(", ") %>);
+void btx_register_matching_callbacks_<%= m.name_sanitized %>(
+    void *btx_handle, <%= m.name_sanitized %>_callback_f *callback);    
+
+<% end %>
```

Comment:
Put it just after the normal callbacks. 
But as usual they should be just one loop for normal-callabcks and matching callbacks.